[FIXED JENKINS-34306] When moving an item several destinations showed in the GUI are not valid

fbelzunc opened the pull request:
1. create a job A in a folder A 
2. create a job with the same name A in another folder B 
3. Then, the move option tells you can move to:

```
Jenkins
Jenkins >> Folder A
Jenkins >> Folder B
````

But this is wrong because `Jenkins >> Folder A` is the item itself and in `Jenkins >> Folder B` there is already an item with the same name - we should cover these cases.

https://issues.jenkins-ci.org/browse/JENKINS-34306

@reviewbybees CC @jglick

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

christ66 commented:
:bug: You can't move into subfolders anymore:

If you have a project in B and create a subfolder:
![57](https://cloud.githubusercontent.com/assets/1302212/14611072/80941a9e-055e-11e6-93af-0d944496a141.png)

The subfolder does not become an available option anymore:
![57-2](https://cloud.githubusercontent.com/assets/1302212/14611076/8345c422-055e-11e6-98bb-de42442814fd.png)

fbelzunc commented:
You are right!

I am fixing it.

fbelzunc commented:
@christ66 Would you mind to review again, please?

christ66 commented:
Tested locally and I was able to move into sub-folders again. :bee:

lordofthejars commented on src/main/java/com/cloudbees/hudson/plugins/folder/relocate/StandardHandler.java:
```diff
@@ -85,13 +86,18 @@ public HandlingMode applicability(Item item) {
-                    Item i = (Item) p;
-                    if (i == item) {
-                        // Cannot move a folder into itself or a descendant.
+                Item i = (Item) p;
```
is normal to not having instanceof `Item` in this case?

jtnord commented:
I think there is a :bug: where it will [always say you can move the item to the root of the Jenkins](https://github.com/jenkinsci/cloudbees-folder-plugin/pull/57/files#diff-800a78b099f2e98dd244ebe63eebd391R80) instance even if there is a job with he same name already existing.

fbelzunc commented:
@jtnord You are right. A new test I created demonstrate you are right.

jglick commented:
@stephenc could say why, but it was intentional that the _current_ position was shown in the pulldown, as the initial selection.

Whether we should quietly hide a folder with an item of the same name, I am not sure. Probably better handled by letting you select that folder but then showing an explicit error.

The original code also blocked you from moving `d1/` into, say, `d1/d2/d3/`, which is a defense you seem to have lost.

fbelzunc commented:
> @stephenc could say why, but it was intentional that the current position was shown in the pulldown, as the initial selection.

I pinged him to understand why it was intended.

> Whether we should quietly hide a folder with an item of the same name, I am not sure. Probably better handled by letting you select that folder but then showing an explicit error.

From User's point of view my opinion is not to expose current position in the pulldown as it will not work - at least for this feature.

> Probably better handled by letting you select that folder but then showing an explicit error.

If something is an error, it shouldn't be exposed unless there is a good reason - waiting @stephenc to answer this. Mainly, because JENKINS-19584 will not block to actually perform the move operation.

> The original code also blocked you from moving d1/ into, say, d1/d2/d3/, which is a defense you seem to have lost.

This was corrected - I added a bunch of tests to add more robustness to this plugin.

stephenc commented:
> I pinged him to understand why it was intended.
 so that by default the move is a no-op if you hit it by mistake

fbelzunc commented:
I always need to read @stephenc 's comments 10 times to understand what he wants to tell me ;-) - so If I understand correctly there is not any good reason to keep the old default behaviour.

stephenc commented:
I think it is a good reason that the default does not move, just like if you open the config page and hit save, by default that will be a no-op unless you change something on the page

fbelzunc commented:
> I think it is a good reason that the default does not move, just like if you open the config page and hit save, by default that will be a no-op unless you change something on the page

Now I understood you ;-)

@jglick @stephenc @christ66 @jtnord  I think I have addressed all the comments now.

jtnord commented:
:bee:

stephenc commented:
:bee:

christ66 commented:
:bee:

recena commented on src/main/java/com/cloudbees/hudson/plugins/folder/relocate/StandardHandler.java:
```diff
@@ -75,7 +76,10 @@ public HandlingMode applicability(Item item) {
+        // ROOT context is only added in case there is not any item with the same name
+        // But we add it in case the one is there is the item itself and not a different job with the same name
+        // No-op by default
+        if (permitted(item, instance) && (instance.getItem(item.getName()) == null) || instance.getItem(item.getName())==item) {
```
`instance.getItem(item.getName()) == item`

fbelzunc replied:
Good catch ;-)

recena commented:
@fbelzunc I'm going to do some tests and LGTM. I'll merge it.

fbelzunc commented:
@recena I already addressed your comment. Would you mind to merge? - I am preparing a new easy PR for this plugin so would like to work with latest bits.

recena commented:
@fbelzunc I recommend you to use the Jira issue ID in you commits `-m "[JENKINS-12345] my message"` or to squash your PR. It is easier when you prepare the changelog for the release notes.

fbelzunc commented:
ok, thanks. I will do next time.
